File: CSharp/OpenTraceability/Models/Events/EventSource.cs
using OpenTraceability.Utility;
using OpenTraceability.Utility.Attributes;
using System.ComponentModel;

namespace OpenTraceability.Models.Events
{
    public enum EventSourceType
    {
        Unknown = 0,

        [CBV("owning_party")]
        [CBV("https://ref.gs1.org/cbv/SDT-owning_party")]
        [CBV("urn:epcglobal:cbv:sdt:owning_party")]
        Owner = 1,

        [CBV("possessing_party")]
        [CBV("https://ref.gs1.org/cbv/SDT-possessing_party")]
        [CBV("urn:epcglobal:cbv:sdt:possessing_party")]
        Possessor = 2,

        [CBV("location")]
        [CBV("https://ref.gs1.org/cbv/SDT-location")]
        [CBV("urn:epcglobal:cbv:sdt:location")]
        Location = 3
    }

    public class EventSource
    {
        [OpenTraceabilityJson("type")]
        [OpenTraceability("@type")]
        public Uri? Type { get; set; }

        public EventSourceType ParsedType
        {
            get
            {
                EventSourceType type = EventSourceType.Unknown;

                foreach (var e in Enum.GetValues<EventSourceType>())
                {
                    if (EnumUtil.GetEnumAttributes<CBVAttribute>(e).Exists(e => e.Value.ToLower() == Type?.ToString().ToLower()))
                    {
                        return e;
                    }
                }

                return type;
            }
            set
            {
                string? t = EnumUtil.GetEnumAttributes<CBVAttribute>(value).Where(e => e.Value.StartsWith("urn")).FirstOrDefault()?.Value;
                if (!string.IsNullOrWhiteSpace(t))
                {
                    this.Type = new Uri(t);
                }
            }
        }

        [OpenTraceabilityJson("source")]
        [OpenTraceability("text()")]
        public string? Value { get; set; }

        public EventSource()
        {

        }
    }
}

File: JDK11/opentraceability/src/main/java/opentraceability/models/events/eventsource.java
package opentraceability.models.events;

import java.net.URI;
import java.util.EnumSet;

import opentraceability.utility.CBVAttribute;
import opentraceability.utility.EnumUtil;
import opentraceability.utility.attributes.OpenTraceabilityAttribute;
import opentraceability.utility.attributes.OpenTraceabilityJsonAttribute;

public class EventSource {
    @OpenTraceabilityJsonAttribute(name = "type")
    @OpenTraceabilityAttribute(ns = "", name = "@type")
    public URI type = null;

    @OpenTraceabilityJsonAttribute(name = "source")
    @OpenTraceabilityAttribute(ns = "", name = "text()")
    public String value = null;

    public EventSourceType getParsedType() {
        EventSourceType type = EventSourceType.Unknown;
        for (EventSourceType e : EnumSet.allOf(EventSourceType.class)) {
            EventSourceType finalType = type;
            if (EnumUtil.GetEnumAttributes(e, CBVAttribute.class).stream().anyMatch(it -> it.value.toLowerCase().equals(finalType.toString().toLowerCase()))) {
                type = e;
                break;
            }
        }

        return type;
    }

    public void setParsedType(EventSourceType value) {
        String t = EnumUtil.GetEnumAttributes(value, CBVAttribute.class).stream().filter(it -> it.value.startsWith("urn")).findFirst().get().value;

        if (t != null && !t.isEmpty()) {
            this.Type = URI.create(t);
        }
    }
}